[#6] <chore>: data.sql 파일 추가

shinsanghooon opened the pull request:
1. 기존에 사용하던 도커 초기화 스크립트인 init.sql을 삭제
2. 데이터 관련 초기화 작업을 더 유연하게 관리하기 위해 data.sql을 사용하도록 변경

Close [#6]

shinsanghooon committed:
[#6] <chore>: data.sql 파일 추가

1. 기존에 사용하던 도커 초기화 스크립트인 init.sql을 삭제
2. 데이터 관련 초기화 작업을 더 유연하게 관리하기 위해 data.sql을
   사용하도록 변경

Close [#6]

kyucumber commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,38 @@
+    phone      VARCHAR(20)  NOT NULL,
+    email      VARCHAR(50)  NOT NULL,
+    ev_car_id     int,
+    foreign key (ev_car_id) references ev_cars (id),
+    card_id    int,
+    foreign key (card_id) references cards (id),
```
최근에는 Foreign Key를 별도로 지정하지 않고 설계하는 경우가 많은 것 같아요.

성능 이슈나 개발할 때 불편하다는 이유 등 다양한 이유가 있을 것 같은데 관련해서 한번 찾아보셔도 좋을 것 같습니다.

Foreign key를 지정하는게 나쁘니 제거하라는 이야기는 아니에요 관련해서 알아두시면 좋을 것 같아서 코멘트 남겨요.

shinsanghooon replied:
FK를 사용하게 되면 insert, update, delete 같은 작업을 할 때 항상 부모 테이블(?)에 값이 있어야 하기 때문에 불편한 점이 있었습니다. 지금 리뷰해주신 `data.sql`에도 `ev_cars`, `cards`, `users` 순서로 테이블을 만든 것도 FK 때문이네요.

성능 이슈는 처음 들어보는거라서 조금 더 찾아봤는데 외래키를 사용하게 되면
- 무결성을 지키기 위해 데이터의 변경이 있을 때마다 동일 데이터의 유무 체크 하는 작업이 추가로 발생하게 됩니다.
- 이 과정에서 테이블에 Lock이 걸리게 되고 이로 인해 대기 시간이 소요됩니다. 
- (외래키도 인덱스가 자동으로 생성 이유도 위와 같은 이유로 조회가 빈번히 발생해서 그런 것 같습니다.) 

대용량이 아닌 경우에는 성능적인 부분에 크게 차이가 없다고 하던데 멘토님은 FK를 지정하지 않고 개발을 하시기도 하는지 궁금합니다.

f-lab-kyucumber replied:
DBA분들이 FK를 사용하지 않는 방향으로 가이드를 주시기도 했고 위와 같은 여러 단점 때문에 저는 FK를 지정하지 않고 애플리케이션 레벨에서 정합성을 맞춰주고 있습니다.

kyucumber commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,38 @@
+    name       VARCHAR(20)  NOT NULL,
+    password   VARCHAR(255) NOT NULL,
+    phone      VARCHAR(20)  NOT NULL,
+    email      VARCHAR(50)  NOT NULL,
```
회원가입 후 로그인 할 때 이메일과 패스워드를 쓸 것 같은데요, 만약 이메일을 통해 로그인한다고 하면 이메일을 통한 조회가 빈번하게 발생할 것 같네요 인덱스가 필요하지 않을까요?
그리고 추가로 유니크 제약 조건도 필요할 것 같습니다.

이 때 유니크 제약 조건과 인덱스, 두개를 모두 설정해야 할까요? 
이미 알고 계실수도 있을 것 같은데 유니크 제약 조건과 인덱스에 대해 한번 찾아보시고 적용해보세요.

shinsanghooon replied:
PK나 UNIQUE 제약 조건이 걸린 컬럼의 경우 인덱스가 자동으로 생성되는 것으로 알고 있습니다. PK나 UNIQUE 제약 조건이 걸린 컬럼은 인덱스로 선택되기 좋은 기준 중 하나인  높은 카디널리티를 만족하고 있어서 자동으로 인덱스가 생성되는 것 같습니다.

kyucumber commented on src/main/resources/data.sql:
```diff
@@ -0,0 +1,38 @@
+    id          INT PRIMARY KEY AUTO_INCREMENT,
+    card_number VARCHAR(30),
+    cvc         VARCHAR(10),
+    yy          VARCHAR(2),
+    mm          VARCHAR(2),
```
이거 두개는 유효기간을 표시하는 것 같은데 yy, mm으로만 표시해도 유추해서 알 순 있겠지만 유효기간임이 조금 더 명확히 표현되면 좋을 것 같아요.

shinsanghooon replied:
yy, mm ... 너무 단순하게 표현한 것 같네요. expiry_year, expiry_month 로 수정하였습니다.

kyucumber commented on src/main/resources/application.yml:
```diff
@@ -16,7 +16,7 @@ spring:
   sql:
     init:
-      mode: never
+      mode: always
```
기존 방법보다는 이게 좀 더 스키마를 관리하기 용이한 것 같습니다. 👍 

이후에 여력이 되시면 flyway 혹은 liquibase를 사용해보셔도 좋을 것 같긴 해요.

shinsanghooon replied:
예전에 flyway를 잠깐 알아본 적이 있는데, 제대로 적용을 못했던 기억이 있네요. 기회가 된다면 다시 도전해보겠습니다 ㅎㅎ

shinsanghooon committed:
chore [#6]: data.sql 수정

1. users 테이블의 email 컬럼은 로그인 시 아이디로 사용되며 조회가 빈번하게 발생할 것으로 예상되어 유니크 제약 조건 추가
2. 모든 테이블의 is_deleted 컬럼은 데이터 생성할 때에는 삭제되는 않은
   경우이므로 디폴트 값을 FALSE로 고정
